packages/otest/lib/Assertion.ts

```diff
@@ -0,0 +1,186 @@
+import { TestError, TestResult } from "./TestResult.js"
+
+export type AssertionDescriber = (description: string) => void
+const noop = () => {}
+
+class AssertionError extends Error {}
+
+interface Class<T> {
+	new (...args: any[]): T
+}
+
+function xor(a: boolean, b: boolean): boolean {
+	return (a && !b) || (b && !a)
+}
+
+function isArguments(a: any) {
+	if ("callee" in a) {
+		for (let i in a) if (i === "callee") return false
+
+		return true
+	}
+}
+
+function testError(errorDescription: string | ErrorConstructor | Class<any>, e: unknown): boolean {
+	if (e == null) return false
+	// make ts shut up, we know what we are doing here, we are ✨ professionals ✨ here
+	const erased = e as any
+	return (
+		(typeof errorDescription === "string" && typeof erased.message === "string" && erased.message === errorDescription) ||
+		e instanceof (errorDescription as any)
+	)
+}
+
+export class Assertion<T> {
+	constructor(private readonly actual: T, private readonly testResult: TestResult) {}
+
+	deepEquals(expected: T): AssertionDescriber {
+		if (!deepEqual(this.actual, expected)) {
+			return this.addError(`expected "${this.actual}" to be deep equal to "${expected}"`)
+		}
+		return noop
+	}
+
+	equals(expected: T | null | undefined): AssertionDescriber {
+		if (this.actual !== expected) {
+			return this.addError(`expected "${this.actual}" to be equal to "${expected}"`)
+		}
+		return noop
+	}
+
+	notDeepEquals(value: T): AssertionDescriber {
+		if (deepEqual(this.actual, value)) {
+			return this.addError(`expected to "${this.actual}" NOT deep equal to "${value}"`)
+		}
+		return noop
+	}
+
+	notEquals(value: T | null | undefined): AssertionDescriber {
+		if (this.actual === value) {
+			return this.addError(`expected "${this.actual}" to NOT be equal to "${value}"`)
+		}
+		return noop
+	}
+
+	satisfies(check: (value: T) => { pass: boolean; message: string }) {
+		const result = check(this.actual)
+		if (!result.pass) {
+			return this.addError(`expected "${this.actual}" to satisfy condition: "${result.message}"`)
+		}
+		return noop
+	}
+
+	async asyncSatisfies(check: (value: T) => Promise<{ pass: boolean; message: string }>) {
```

Comment:
what would prevent us from merging `satisfies` and `asyncSatisfies` to simplify the API?